Bottom: 684e21cf3e81b4f5bd740456e0fea103e52d7d1f
Top:    729b947f8c52c96a11c4a33ff4ef447cd22cc5d1
Author: Sean Hefty <sean.hefty@intel.com>
Date:   2012-06-04 13:22:10 -0700

rsocket: Handle TCP_MAXSEG socket option

netperf uses the TCP_MAXSEG socket option.  Add support for it.
Problem reported by Sridhar Samudrala <sri@us.ibm.com>

Signed-off-by: Sean Hefty <sean.hefty@intel.com>


---

diff --git a/src/rsocket.c b/src/rsocket.c
index a1679f1..e899e8a 100644
--- a/src/rsocket.c
+++ b/src/rsocket.c
@@ -1717,6 +1717,9 @@ int rsetsockopt(int socket, int level, int optname,
 			opt_on = *(int *) optval;
 			ret = 0;
 			break;
+		case TCP_MAXSEG:
+			ret = 0;
+			break;
 		default:
 			break;
 		}
@@ -1814,6 +1817,12 @@ int rgetsockopt(int socket, int level, int optname,
 			*((int *) optval) = !!(rs->tcp_opts & (1 << optname));
 			*optlen = sizeof(int);
 			break;
+		case TCP_MAXSEG:
+			*((int *) optval) = (rs->cm_id && rs->cm_id->route.num_paths) ?
+					    1 << (7 + rs->cm_id->route.path_rec->mtu) :
+					    2048;
+			*optlen = sizeof(int);
+			break;
 		default:
 			ret = ENOTSUP;
 			break;
